### ISSUE TITLE
G2 block top toolbar

### ISSUE THREAD

**Issue by pablohoneyhoney** [WordPress/gutenberg#20592]
Here a few explorations on how to visualize the top toolbar when active, adapting [G2](https://github.com/WordPress/gutenberg/issues/18667) styles. 
Some alternatives permit the top bar to remain clean to potentially absorb more functionality and stay focused. 

<img width="1440" alt="Top toolbar" src="https://user-images.githubusercontent.com/25516611/75699757-f4263200-5c7e-11ea-941e-880d42830757.png">

**A (Absorbed by top bar as is today)**

<img width="1440" alt="Top toolbar-6" src="https://user-images.githubusercontent.com/25516611/75699965-5717c900-5c7f-11ea-8f92-5e4e48c46904.png">

**B (Creating a second level)**

<img width="1440" alt="Top toolbar-5" src="https://user-images.githubusercontent.com/25516611/75699976-5c751380-5c7f-11ea-87c2-666163b5c9ee.png">

**C (Aligned with the top bar UI, 2 options)**

<img width="1440" alt="Top toolbar-4" src="https://user-images.githubusercontent.com/25516611/75699986-60a13100-5c7f-11ea-8453-301281321093.png">

<img width="1440" alt="Top toolbar-1" src="https://user-images.githubusercontent.com/25516611/75700270-cc839980-5c7f-11ea-9f79-455edac4ae25.png">

**D (Horizontal position fixed pointing to the block content area)**

<img width="1440" alt="Top toolbar-3" src="https://user-images.githubusercontent.com/25516611/75700003-68f96c00-5c7f-11ea-87da-9939f7014da5.png">

**E (Centered)**

<img width="1440" alt="Top toolbar-2" src="https://user-images.githubusercontent.com/25516611/75700117-8c241b80-5c7f-11ea-9d32-2154b9f73ec2.png">

Relevant to:
https://github.com/WordPress/gutenberg/issues/20575
https://github.com/WordPress/gutenberg/issues/6047
https://github.com/WordPress/gutenberg/issues/7485
https://github.com/WordPress/gutenberg/issues/3735

**Comment by shaunandrews** [comment]
Options C, D, and E could cause trouble by obscuring content within the canvas — especially considering full-site editing; ;The toolbar could obscure things like a header, logo, navigation, etc.

I have a slight preference for B, over A; A feels a little too heavy (with extra borders) and isn't realistic on mobile screens. I _believe_ B is how the toolbar works on mobile today.

**Comment by shaunandrews** [comment]
Related to obscuring content, we could just add extra whitespace at the top of the canvas. This is something I've explore in the context of editing a template, where its likely that a `template_part` will exist at the top of the canvas.

![Push Header Down](https://user-images.githubusercontent.com/191598/75718133-68250200-5ca0-11ea-81d9-8a64b3960dbe.gif)

However, in the context of the Top Toolbar mode, I'd expect the whitespace to always display.

**Comment by karmatosed** [comment]
My initial preference is absorbing into top navigation, A. In saying that there is something about option C, the first visual. I know in testing there has been almost a 50/50 split in those that want 'by block' and those that want in the toolbar. I almost feel we might need to keep that option in future, or at least phase out gradually. I say this knowing I am suggesting a choice, however, the default might end up being the one people stay with.

I am nodding to your comment about the canvas @shaunandrews and that makes me even more keen on the absorbed option. Whilst the 'bounce in' isn't too dramatic it still feels like it takes me out of the experience and further makes it not feel as would on the front. That might be not an issue though if doesn't fracture the interaction.

**Comment by jasmussen** [comment]
I love seeing these varying mockups even for things we thought were "stable" and decided upon. A and B is basically what we have today, but with a few visual flourishes perhaps, and after seeing the alternate options, that's still what feels the most right for people who explicitly choose the top toolbar option.

However I like the other options more than I thought I would, and even just exploring them has inspirational value!

**Comment by ZebulanStanphill** [comment]
With the changes proposed in #20877 in mind, I prefer options C and D.

I don't like the current behavior (A) since it merges one toolbar into the middle of another one, and it makes things like #20877 more difficult to implement.

B feels unnecessarily heavy, and considering @shaunandrews comment, I don't think there's anything to worry about regarding obscuring content.

I'm not too big a fan of E since it makes the position of the block icon jump around as you select different blocks. The point of the "Top toolbar" option is to keep the toolbar in a single place, and so anchoring the left side to a specific point as in C and D seems like the right way to go.

I should note that I never use the "Top toolbar" option, so I'm probably not the best person to ask regarding how it should work, since I prefer the block-anchored toolbars anyway.

**Comment by paaljoachim** [comment]
Btw here is the top toolbar on a smaller resized desktop screen.

What I like about todays method is that it shows the toolbar below the regular top bar area. 
1 row - standard top bar related icons.
2 row - the block toolbar.
It just keeps a clear difference between the two.

![Screen Shot 2020-07-07 at 18 31 13](https://user-images.githubusercontent.com/5323259/86813726-d0398980-c080-11ea-9789-78f0fd25d572.png)

**Comment by estelaris** [comment]
As discussed today in the design triage, the team prefers option B as it seems pretty standard and a method used by other software. The question is if that overflow row will it pre-exist and then be filled out by the icons from the secondary menu? or will it "bounce in" (not preferred method)?

**Comment by ZebulanStanphill** [comment]
The G2 toolbar is no longer trying to stick out a bit to the left of the current block, and the little black triangle thing is gone, so I think option D is off the table now.

I know I said I prefer option C, but having thought about it more, I've changed my mind: **I now prefer option B**. It's a lot more clear that the toolbar is stuck to the top in that design than in C, D, or E. The only thing I'm not certain about is the exact horizontal alignment, but that's just a minor detail.

**Comment by ZebulanStanphill** [comment]
Also, the "heaviness" problem could be solved by just changing the border to look the same as it already does at medium viewport sizes, as seen in @paaljoachim's screenshot.

**Comment by OlaIola** [comment]
This absorbed current version confused me. I was sure that something is wrong. I was needed to move a block around and make it from a keyboard is very difficult, I was sure that it isn't working also. And with this menu stuck to the top isn't possible to drag block by it.
This menu needs a way to unstick at the same point where it stands.

**Comment by ZebulanStanphill** [comment]
I've been thinking about this some more, and I think @paaljoachim's mockup is _exactly_ what we should do. Trying to horizontally align the block toolbar to the current block is pointless when you're not going to show it right next to the block, and would be confusing in the context of horizontally-laid-out blocks anyway.

Pinging @afercia to check this out since I don't think he's seen it yet. I think this change could be a benefit to editor a11y, since it would keep the block-level toolbar visually separated from the document-level top bar.

**Comment by afercia** [comment]
A clearer separation between the document-level toolbar and the block-level toolbar could help all users.

Amongst the option above, B seems to me the most appropriate also because it's basically what we have now for medium/small viewports. I'd leave considerations on whether a split toolbar respect the original intent of the "Top toolbar" to others.

As usual, when it comes to accessibility it's not just a matter of visuals though. Semantics and keyboard interaction are of fundamental importance. Right now, even if the toolbar goes in two lines it's still wrapped in a single element with role=toolbar and aria-orientation=horizontal:

<img width="1026" alt="Screenshot 2020-09-29 at 14 49 58" src="https://user-images.githubusercontent.com/1682452/94562706-f167db80-0265-11eb-8478-431dd507e901.png">

Arrows navigation works for all the buttons within the area highlighted in red in the screenshot above. Which feels weird for keyboard users s they would expect keyboard navigation match the _visual order_.

Also, if we want a clearer _visual_ separation then we should represent this separation _semantically_. I'd recommend to separate the two toolbars and wrap them within _two_ elements with role=toolbar. Arrows navigation within the two toolbars should work separately so that the tab order is:
- tab -> document toolbar (one single tab stop)
- tab -> Preview -> tab -> Publish -> tab -> Settings -> More menu
- tab -> block toolbar (one single tab stop)

Having two separate role=toolbar elements would also allow to better label the toolbar. Right now, the single top bar is labelled with `aria-label="Document and block tools"`. With two separate toolbars we could use two labels:
- aria-label="Document tools"
- aria-label="Block tools"

**Jumping to the toolbar(s) with Alt+F10**
Right now, with the unified toolbar, it is possible to move focus and jump from the selected block to the toolbar by pressing the keys Alt+F10. Note: this key combination comes from the Classic Editor and it was kept because users are familiar with it. 

With the unified toolbar this is not a problem because all controls are grouped together. WIth the two split toolbars, where focus should be moved to when pressing Alt+F10? To the document-level toolbar or to the block-level one?

**Comment by talldan** [comment]
Marking this issue as stale, as it seems quite outdated.

It might be worth making smaller individual actionable issues if there's still anything to do.
